From f2f90d10d1411f9eee1019997174acda2e19daf2 Mon Sep 17 00:00:00 2001
From: Suman Anna <s-anna@ti.com>
Date: Fri, 15 Feb 2019 12:38:35 -0600
Subject: [PATCH 37/51] soc: ti: pruss: Add support for PRU-ICSS subsystems on
 66AK2G SoC

commit eb3ba131259b56933e67c189af466aada88af003 from
https://git.ti.com/cgit/processor-sdk/processor-sdk-linux/?h=processor-sdk-linux-01.00.00

The 66AK2G SoC supports two PRU-ICSS instances, named PRUSS0 and PRUSS1,
each of which has two PRU processor cores. The two PRU-ICSS instances
are identical to each other with few minor SoC integration differences,
and are very similar to the PRU-ICSS1 of AM57xx/AM43xx. The Shared Data
RAM size is larger and the number of interrupts coming into MPU INTC
is like the instances on AM437x. There are also few other differences
attributing to integration in Keystone architecture (like no SYSCFG
register or PRCM handshake protocols). Other IP level differences
include different constant table, differences in system event interrupt
input sources etc. They also do not have a programmable module reset
line like those present on AM33xx/AM43xx SoCs. The modules are reset
just like any other IP with the SoC's global cold/warm resets.

The existing PRUSS platform drivers have been enhanced to support
these 66AK2G PRU-ICSS instances through new 66AK2G specific compatibles
for properly probing and booting all the different PRU cores in each
PRU-ICSS processor subsystem. The SoC-level integration differences
are dealt with using match data within the PRUSS INTC and PRUSS SoC
bus drivers. A build dependency with ARCH_KEYSTONE is added to enable
the driver to be built in K2G-only configuration. The initial names
for the firmware images for each PRU core are retrieved from DT nodes,
and can be adjusted through sysfs if required.

Signed-off-by: Andrew F. Davis <afd@ti.com>
Signed-off-by: Suman Anna <s-anna@ti.com>
Signed-off-by: Xiaolei Wang <xiaolei.wang@windriver.com>
---
 drivers/irqchip/irq-pruss-intc.c | 8 ++++++++
 drivers/remoteproc/pru_rproc.c   | 1 +
 drivers/soc/ti/Kconfig           | 2 +-
 drivers/soc/ti/pruss.c           | 1 +
 drivers/soc/ti/pruss_soc_bus.c   | 4 ++++
 5 files changed, 15 insertions(+), 1 deletion(-)

diff --git a/drivers/irqchip/irq-pruss-intc.c b/drivers/irqchip/irq-pruss-intc.c
index ee25959be195..5885ccb44423 100644
--- a/drivers/irqchip/irq-pruss-intc.c
+++ b/drivers/irqchip/irq-pruss-intc.c
@@ -600,6 +600,10 @@ static const struct pruss_intc_match_data am437x_pruss_intc_data = {
 	.no_host7_intr = true,
 };
 
+static const struct pruss_intc_match_data k2g_pruss_intc_data = {
+	.no_host7_intr = true,
+};
+
 static const struct of_device_id pruss_intc_of_match[] = {
 	{
 		.compatible = "ti,am3356-pruss-intc",
@@ -609,6 +613,10 @@ static const struct of_device_id pruss_intc_of_match[] = {
 		.compatible = "ti,am4376-pruss-intc",
 		.data = &am437x_pruss_intc_data,
 	},
+	{
+		.compatible = "ti,k2g-pruss-intc",
+		.data = &k2g_pruss_intc_data,
+	},
 	{ /* sentinel */ },
 };
 MODULE_DEVICE_TABLE(of, pruss_intc_of_match);
diff --git a/drivers/remoteproc/pru_rproc.c b/drivers/remoteproc/pru_rproc.c
index 6551dc790b55..8339ca960e13 100644
--- a/drivers/remoteproc/pru_rproc.c
+++ b/drivers/remoteproc/pru_rproc.c
@@ -594,6 +594,7 @@ static int pru_rproc_remove(struct platform_device *pdev)
 static const struct of_device_id pru_rproc_match[] = {
 	{ .compatible = "ti,am3356-pru", },
 	{ .compatible = "ti,am4376-pru", },
+	{ .compatible = "ti,k2g-pru",    },
 	{},
 };
 MODULE_DEVICE_TABLE(of, pru_rproc_match);
diff --git a/drivers/soc/ti/Kconfig b/drivers/soc/ti/Kconfig
index 51f9d0caa90d..89bb4d8acb73 100644
--- a/drivers/soc/ti/Kconfig
+++ b/drivers/soc/ti/Kconfig
@@ -77,7 +77,7 @@ config TI_SCI_PM_DOMAINS
 
 config TI_PRUSS
 	tristate "TI PRU-ICSS Subsystem Platform drivers"
-	depends on SOC_AM33XX || SOC_AM43XX
+	depends on SOC_AM33XX || SOC_AM43XX || SOC_DRA7XX || ARCH_KEYSTONE
 	select MFD_SYSCON
 	help
 	  TI PRU-ICSS Subsystem platform specific support.
diff --git a/drivers/soc/ti/pruss.c b/drivers/soc/ti/pruss.c
index b88c1d596a93..51ef9c4746d0 100644
--- a/drivers/soc/ti/pruss.c
+++ b/drivers/soc/ti/pruss.c
@@ -202,6 +202,7 @@ static const struct pruss_match_private_data am437x_match_data[] = {
 static const struct of_device_id pruss_of_match[] = {
 	{ .compatible = "ti,am3356-pruss", .data = NULL, },
 	{ .compatible = "ti,am4376-pruss", .data = &am437x_match_data, },
+	{ .compatible = "ti,k2g-pruss", .data = NULL, },
 	{ /* sentinel */ },
 };
 MODULE_DEVICE_TABLE(of, pruss_of_match);
diff --git a/drivers/soc/ti/pruss_soc_bus.c b/drivers/soc/ti/pruss_soc_bus.c
index d06fe9c091da..1b9f6cb5d830 100644
--- a/drivers/soc/ti/pruss_soc_bus.c
+++ b/drivers/soc/ti/pruss_soc_bus.c
@@ -20,10 +20,12 @@
  * struct pruss_soc_bus - PRUSS SoC bus structure
  * @syscfg: kernel mapped address for SYSCFG register
  * @has_reset: cached variable for storing global module reset flag
+ * @skip_syscfg: flag to indicate if PRCM master standby/slave idle is needed
  */
 struct pruss_soc_bus {
 	void __iomem *syscfg;
 	bool has_reset;
+	bool skip_syscfg;
 };
 
 /**
@@ -32,6 +34,7 @@ struct pruss_soc_bus {
  */
 struct pruss_soc_bus_match_data {
 	bool has_reset;
+	bool uses_prcm;
 };
 
 static int pruss_soc_bus_probe(struct platform_device *pdev)
@@ -63,6 +66,7 @@ static int pruss_soc_bus_probe(struct platform_device *pdev)
 		return -ENODEV;
 	}
 	psoc_bus->has_reset = data->has_reset;
+	psoc_bus->skip_syscfg = !data->uses_prcm;
 	platform_set_drvdata(pdev, psoc_bus);
 
 	if (psoc_bus->has_reset) {
-- 
2.17.1

